### PR TITLE
Allow multiple about:config settings per option

### PR DESCRIPTION
**Pull request by allo-**

### CODE REVIEW

File: prefs.json [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
   "prefs": {
-    "dom.event.clipboardevents.enabled": {
+    "clipboardevents": {
       "categories": [ "html5-api-dom", "data-leak" ],
       "type": "boolean",
       "description": "Disable clipboard event detection (onCut/onCopy/onPaste) via Javascript",
-      "value_when_active": false,
+      "settings": [
+          "dom.event.clipboardevents.enabled": false
+      ]
```
Comment (nodiscc) [inline prefs.json]: Missing `,`

File: prefs.json [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
     },
-    "network.http.referer.spoofSource": {
+    "referer_spoof_source": {
       "categories": [ "http", "tracking" ],
       "type": "boolean",
       "description": "Send a referer header with the target URI as the source",
-      "value_when_active": true,
+      "settings": [
+          "network.http.referer.spoofSource": true
+      ]
```
Comment (nodiscc) [inline prefs.json]: Missing `,`

File: prefs.json [hunk 3]
```diff
@@ -27,23 +31,27 @@
...
     },
-    "browser.cache.disk.capacity": {
+    "disk_cache": {
       "categories": [ "disk-caching", "performance" ],
       "type": "integer",
       "description": "Disable the disk cache to reduce browsing data on your pc",
-      "value_when_active": 0,
+      "settings": [
+          "browser.cache.disk.capacity": 0
+      ]
```
Comment (nodiscc) [inline prefs.json]: Missing `,`

File: prefs.json [hunk 4]
```diff
@@ -27,23 +31,27 @@
...
-    "general.useragent.override": {
+    "useragent": {
       "categories": [ "tracking", "category3" ],
       "type": "integer",
-      "description": "Change your useragent string",
-      "value_when_active": "USERINPUT",
+      "description": "Change your useragent string. If you keep the field empty, no setting will be changed.",
+      "settings": [
+          "general.useragent.override": ""
+      ]
```
Comment (nodiscc) [inline prefs.json]: Missing `,`

**Review by nodiscc — DISMISSED** [review]
Multiple missing commas;

can you also add an example entry that would control multiple preferences? So that we get a real example to work on.

**Comment by allo-** [comment]
Okay, now we're at the option types. This works for a checkbox option, but i.e. the ``general.useragent.override`` setting is wrong this way. Either it must be changed to a checkbox setting (with a static fake useragent) or the handling is different.

I had three types of options:
- checkbox (sets one or more values to given values like we had in ``value_when_active``)
- choice box (like checkbox, but with one set of values (and possible settings) per choice)
- textbox (sets exactly one setting to the given string)

**Comment by allo-** [comment]
You need to add your changes yourself, i cannot approve something in my own pullrequest.

**Comment by nodiscc** [comment]
> Okay, now we're at the option types.

Let's stick to working on multiple preferences per option for this Pull Request.

``` 
    "disk_cache": {
       "categories": [ "disk-caching", "performance" ],
       "type": "integer",
       "description": "Disable the disk cache to reduce browsing data on your pc",
      "settings": {
          "browser.cache.disk.capacity": 0
      }, 
```

The problem with that is: some profile/user.js maintainers will want to disable disk caching with `"browser.cache.disk.capacity": 0`, while some [others](https://github.com/pyllyukko/user.js/blob/master/user.js#L730) will prefer `"browser.cache.disk.enable": false`. How do you handle that case?

**Comment by allo-** [comment]
For multiple settings we first need to change the semantics from preferences (bool, string, integer) to options (checkbox, string input, numeric string (or even number input in html5)).
Then it becomes obvious, that string and number settings are always one-to-one.

> The problem with that is: some profile/user.js maintainers will want to disable disk caching with "browser.cache.disk.capacity": 0, while some others will prefer "browser.cache.disk.enable": false. How do you handle that case?

Two (or three if you want both to be sure) different settings.json files, and each maintainer references the one he likes best. That's why I want to split it in many small snippets, which represent a single feature (not more than one feature at once, but more than one setting if it's needed for the feature).

The more problematic case is, when some maintainer includes overlapping settings files. Then the result is implementation specific. Does it trigger an error message? Does the later one override the prior one? And "set default value when unchanged" vs. "set only when changed" makes a difference.

But i think this won't happen too often and if there are problems the maintainer of a profile needs to split settings and create new settings files which match what he wants to achieve with his set of preferences.

**Comment by nodiscc** [comment]
At this point the schema is almost usable. You seem to have a clearer idea than I do for the remaining details, I have transfered this repo to you. I suggest that you make the remaining changes that would fit your workflow, I will keep an eye on this and https://github.com/Genbox/HardenIT/issues/2 and will try to write a `user.js -> json` converter for the exact format you decide on.

**Comment by allo-** [comment]
Okay.
I guess I will try to match it with my existing files and possibly create a shared repo for collecting settings files. 

It would be cool, when we can motivate as many people as possible to contribute their settings, so we get a big repo which everyone who wants to build firefox profiles can use.

**Pull request merged**